pkg/mpc-examples/contracts/04-pause-unpause-controller/PauseUnpauseController.sol

```diff
@@ -0,0 +1,118 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-interfaces/contracts/pool-utils/ILastCreatedPoolFactory.sol";
+import "@balancer-labs/v2-interfaces/contracts/pool-utils/IManagedPool.sol";
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+contract PauseUnpauseController is Ownable {
+    uint256 private constant _END_SWAP_FEE_PERCENTAGE = 3e15;
+    uint256 private constant _START_SWAP_FEE_PERCENTAGE = 95e16;
+    uint256 private constant _SWAP_FEE_REDUCATION_DURATION = 3 days;
+
+    IVault private _vault;
+    bytes32 private _poolId;
+
+    constructor(IVault vault, address controllerOwner) {
+        // Get poolId from the factory.
+        bytes32 poolId = IManagedPool(ILastCreatedPoolFactory(msg.sender).getLastCreatedPool()).getPoolId();
+
+        // Verify that this is a real Vault and the pool is registered - this call will revert if not.
+        vault.getPool(poolId);
+
+        // store Vault and PoolId.
+        _poolId = poolId;
+        _vault = vault;
+
+        // Transfer ownership from factory to manager.
+        transferOwnership(controllerOwner);
+    }
+
+    /// === public Getters ===
+
+    function getPoolId() public view returns (bytes32) {
+        return _poolId;
+    }
+
+    function getVault() public view returns (IVault) {
+        return _vault;
+    }
+
+    function getSwapEnabled() public view returns (bool) {
+        return _getPool().getSwapEnabled();
+    }
+
+    function isPoolPaused() public view returns (bool) {
+        return !_getPool().getSwapEnabled();
+    }
+
+    /// === Setters ===
+
+    /**
+     * @notice Disables swapping.
+     */
+    function pausePool() external onlyOwner returns (bool) {
+        require(getSwapEnabled(), "swapping with pool is already paused");
+        _getPool().setSwapEnabled(false);
+        // pool is confirmed paused
+        return true;
+    }
+
+    /**
+     * @notice Unpauses the pool in a safe or unsafe manner.
+     * @dev A safe unpause is desirable as the market likely had price movements
+     * which have not been reflected in a paused pool. In order to not leak too much
+     * arbitrage losses, the controller adjusts swap fees of the managed pool to
+     * _START_SWAP_FEE_PERCENTAGE instantly and let's the market bring the pool back
+     * into balance via minimal viable arbitrage.
+     * @param shouldSafeUnpause Decision to safely unpause the pool.
+     */
+
+    /* solhint-disable not-rely-on-time */
+    function unpausePool(bool shouldSafeUnpause) external onlyOwner returns (bool) {
+        if (shouldSafeUnpause) {
```

Comment:
I like the idea of offering safe and unsafe methods here; the word "should" in `shouldSafeUnpause` seems a bit uncertain to me though. How would you feel about changing that name (and therefore flipping the logic a little) to be something like `forceUnsafeUnpause`?

Reply:
Actually maybe it makes sense to break this out into two functions: `safeUnpausePool` and `dangerousUnpausePool`. Then they could both call an internal function that does `_getPool().setSwapEnabled(true);` and the `safeUnpausePool` would just kick off the swap fee ramp before doing so

Reply:
I think using forceUnsafeUnpause is more clear, yes. Rab and I were thinking whether to make this one or two functions. We opted for making it one as it forces the caller to reason about the type of unpause he wants to take. Splitting it up in two functions. However I think the more important consideration is that it is clear for the caller if he is using the safe or unsafe unpause. Proper function naming imo can also ensure that. I will rework it to be two functions.